lib/std/log.zig

```diff
@@ -0,0 +1,81 @@
+const std = @import("std.zig");
+const debug = std.debug;
+const builtin = std.builtin;
+const root = @import("root");
+
+pub const Level = enum {
+    /// This log level is intended to be used for any information about the
+    /// state of the program that would be useful while debugging.
+    info,
+    /// This log level is intended be used when it is uncertain if something
+    /// has gone wrong or not, but the circumstances are worth investigating.
+    warn,
+    /// This log level is intended be used when an error has occurred but the
+    /// program is able to recover and continue running.
+    err,
+    /// This log level should only be used for irrecoverable errors as it will
+    /// trigger a panic after logging. Note that this will not log if the build
+    /// mode is release small but will still trigger a panic.
+    fatal,
+};
+
+/// The default log level is based on build mode. Note that in ReleaseSmall
+/// builds logging is totally disabled (log.fatal still causes a panic though).
+pub const default_level: Level = switch (builtin.mode) {
+    .Debug => .info,
+    .ReleaseSafe => .warn,
+    .ReleaseFast => .err,
+    .ReleaseSmall => .fatal,
+};
+
+/// The current log level. This is set to root.log_level if present, otherwise
+/// log.default_level.
+pub const level: Level = if (@hasDecl(root, "log_level")) root.log_level else default_level;
+
+fn log(comptime message_level: Level, comptime format: []const u8, args: var) void {
```

Comment:
```suggestion
pub fn log(comptime message_level: Level, comptime format: []const u8, args: var) void {
```

Reply:
I believe that making this public would violate the "Only one obvious way to do things" bullet point of the zig zen. I prefer
```zig
std.log.warn("message", .{});
```
over 
```zig
std.log.log(.warn, "message", {});
```
personally so I'm inclined to keep this as is.

Reply:
`pub fn log` will support custom level if `Level` become non-exhaustive enum.

Reply:
Frankly I'm not a fan of allowing custom log levels when using `std.log`. It is imo unnecessary complexity and will reduce compatibility between libraries and programs which is the whole point of a unified `std.log` interface.

Reply:
Lack of flexibility forces developers to reinvent their own wheels.